lib/db/observed.go

```diff
@@ -0,0 +1,217 @@
+				goto deleteKey
+			}
+			if err := of.Unmarshal(bs); err != nil {
+				goto deleteKey
+			}
+			if _, ok := res[folderID]; !ok {
+				res[folderID] = make(map[protocol.DeviceID]ObservedFolder)
+			}
+			res[folderID][deviceID] = of
+			continue
+		}
+	deleteKey:
+		l.Warnln("Invalid pending folder entry, deleting from database")
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+// Set of devices for which pending folders are allowed, but not specific folder IDs
+type DropListObserved map[protocol.DeviceID]map[string]struct{}
+
+func (dl DropListObserved) MarkDevice(device protocol.DeviceID) map[string]struct{} {
+	folders, ok := dl[device]
+	if !ok {
+		dl[device] = nil
+	}
+	return folders
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID) {
+	for _, dev := range devices {
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]struct{})
+			dl[dev] = folders
+		}
+		folders[folder] = struct{}{}
+	}
+}
+
+// CleanPendingDevices removes all pending device entries matching a given set of device IDs
+func (db *Lowlevel) CleanPendingDevices(dropList DropListObserved) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		l.Warnf("Could not iterate through pending device entries for cleanup: %v", err)
+		return
+	}
+	defer iter.Release()
+	for iter.Next() {
+		keyDev := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(keyDev) != protocol.DeviceIDLength {
+			l.Warnf("Invalid pending device entry, deleting from database: %v", keyDev)
+		} else {
+			// Valid entries are looked up in the drop-list, invalid ones cleaned up
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			_, dropDev := dropList[deviceID]
+			if !dropDev {
+				continue
+			}
+			l.Debugf("Removing marked pending device %v", deviceID)
+		}
+		if err := db.Delete(iter.Key()); err != nil {
+			l.Warnf("Failed to remove pending device entry: %v", err)
+		}
+	}
+}
+
+// CleanPendingFolders removes all pending folder entries not matching a given set of
+// device IDs, or matching the set of folder IDs associated with those given devices.
+func (db *Lowlevel) CleanPendingFolders(dropList DropListObserved) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+	if err != nil {
+		l.Warnf("Could not iterate through pending folder entries for cleanup: %v", err)
+		return
+	}
+	defer iter.Release()
+	for iter.Next() {
+		if keyDev, ok := db.keyer.DeviceFromPendingFolderKey(iter.Key()); !ok {
+			l.Warnf("Invalid pending folder entry, deleting from database: %v", keyDev)
+		} else {
+			// Valid entries are looked up in the drop-list, invalid ones cleaned up
+			deviceID := protocol.DeviceIDFromBytes(keyDev)
+			folders, keepDev := dropList[deviceID]
```

Comment:
I think this `keepDevice` just cause me a lot of confusion: Isn't a device present here a device we want to drop from pending devices? Thus either `dropDev` or `haveDev` would make more sense to me.

Reply:
> Isn't a device present here a device we want to drop from pending devices?

Yes, but that's a diffferent context. Within `CleanPendingFolders()` we enumerate devices that are known, so **only** those **can** have pending folders at all. For each of them, there are known and ignored folders, which should **not** be pending anymore. So having the device in that first level map means "keep, except for the following".

I understand that variable name can be slightly confusing. How about `allowDev`?

Reply:
Just that both are enumerated in the same object, so in my head to not confuse things I considered them as "devices we have handled, i.e. in config/ignored/..., but not pending" and `keepDev` didn't fit in well. The new `allowDev` is fine too.  
(This e.g. can now be resolved, as we've come to consensus, and by you, not by me, because I'd like you to read my/this comment - maybe my "review-comment-resolving-model" is a bit obscure, I hope not too much :) ).

Reply:
No it's fine, I was indeed a little trigger-happy this weekend on resolving.  I do read every comment as an e-mail and they stay in the inbox until the conrete topic is dismissed for me.  GitHub's conversation view is really not so great for these review discussions, hopefully they will come up with a better overview sometime.